cmd/duffle/relocate.go

```diff
@@ -0,0 +1,249 @@
+bundle with an updated image map.
+
+The --repository-prefix flag determines the repositories for the relocated images.
+Each image is tagged with a name starting with the given prefix and pushed to the repository.
+
+For example, if the repository-prefix is example.com/user, the image istio/proxyv2 is relocated
+to a name starting with example.com/user/ and pushed to a repository hosted by example.com.
+`
+
+type relocateCmd struct {
+	// args
+	inputBundle  string
+	outputBundle string
+
+	// flags
+	repoPrefix         string
+	inputBundleIsFile  bool
+	outputBundleIsFile bool
+
+	// context
+	home home.Home
+	out  io.Writer
+
+	// dependencies
+	mapping        pathmapping.PathMapping
+	registryClient registry.Client
+}
+
+func newRelocateCmd(w io.Writer) *cobra.Command {
+	relocate := &relocateCmd{out: w}
+
+	cmd := &cobra.Command{
+		Use:   "relocate [INPUT-BUNDLE] [OUTPUT-BUNDLE]",
+		Short: "relocate images in a CNAB bundle",
+		Long:  relocateDesc,
+		Example: `duffle relocate helloworld hellorelocated --repository-prefix example.com/user
+duffle relocate path/to/bundle.json relocatedbundle --repository-prefix example.com/user --input-bundle-is-file
+duffle relocate helloworld path/to/relocatedbundle.json --repository-prefix example.com/user --output-bundle-is-file`,
+		Args: cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// validate --repository-prefix if it is set, otherwise allow flag omission to be diagnosed as such
+			if cmd.Flags().Changed("repository-prefix") {
+				if err := flagValidRepository("repository-prefix")(cmd); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			relocate.inputBundle = args[0]
+			relocate.outputBundle = args[1]
+
+			relocate.home = home.Home(homePath())
+
+			relocate.mapping = pathmapping.FlattenRepoPathPreserveTagDigest
+			relocate.registryClient = registry.NewRegistryClient()
+
+			return relocate.run()
+		},
+	}
+
+	f := cmd.Flags()
+	f.BoolVarP(&relocate.inputBundleIsFile, "input-bundle-is-file", "", false, "Indicates that the input bundle source is a file path")
+	f.BoolVarP(&relocate.outputBundleIsFile, "output-bundle-is-file", "", false, "Indicates that the output bundle destination is a file path")
+	f.StringVarP(&relocate.repoPrefix, "repository-prefix", "r", "", "Prefix for relocated image names")
+	cmd.MarkFlagRequired("repository-prefix")
+
+	return cmd
+}
+
+func (r *relocateCmd) run() error {
+	bun, err := r.setup()
+	if err != nil {
+		return err
+	}
+
+	if err := r.relocate(bun); err != nil {
+		return err
+	}
+
+	if _, err := r.writeBundle(bun); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *relocateCmd) relocate(bun *bundle.Bundle) error {
+	// mutate the input bundle to become the output bundle
+	if !r.outputBundleIsFile {
+		bun.Name = r.outputBundle
+	}
+
+	ks := keys(bun.Images)
+	for _, k := range ks {
```

Comment:
I think it would be simpler to just iterate on the `Images` map and remove this `keys` function:
```go
for k, im := range bun.Images {
    if !isOCI(im.ImageType) && !isDocker(im.ImageType) {
        continue
    }
}
```
You can also "fast return" using a `continue`, but this last point is totally nit picking 👍 .

Reply:
I was nervous about mutating `bun.Images` while iterating over it, but it seems changing map values during the iteration is safe, so I'm adopting your suggestion.

I'm going to skip introducing the "fast return" because:
* I find the code less easy to read
* any speed improvement would be neglible and performance isn't critical here
* I'd expect the Go compiler's optimiser to strip out unnecessary branches, so the current code is probably just as fast.

Reply:
I looked at the resulting assembly code using `go tool objdump -S -s relocate  duffle` and when `isOCI(im.ImageType) || isDocker(im.ImageType)` is `false`, there is a jump straight back to the map iteration. :-)

Reply:
Oh sure, but it wasn't an optimisation for performance, but more a coding style. I should have said "early return" by the way, not "fast return", my bad.